File: itunes.hs
{-# LANGUAGE ViewPatterns        #-}
{-

 itunes

 AUTHOR: Chris Barrett <chris.d.barrett@me.com>
 LICENSE: BSD

 Copyright (c) 2013, Chris Barrett

 DESCRIPTION:
  Commands for working with iTunes from the command-line.

 CABAL DEPENDENCIES:
  ansi-wl-pprint
  zip-archive = 0.1.3.4

-}

{-# LANGUAGE ScopedTypeVariables #-}
import           Codec.Archive.Zip
import           Control.Applicative
import           Control.Exception
import           Control.Monad
import qualified Data.ByteString.Lazy.Char8   as L8
import           Data.Char                    (toLower)
import           Prelude                      hiding (catch)
import           System.Directory
import           System.Environment           (getArgs)
import           System.Exit                  (exitFailure)
import           System.FilePath.Posix
import           Text.PrettyPrint.ANSI.Leijen (dullyellow, green, linebreak,
                                               putDoc, red, text, (<+>), (<>))

--- Enumerates the possible parsed values of the program arguments.
data Args = Add [FilePath] | Help | Invalid | Unknown String
          deriving Show

--- Enumerates different filetypes that need to be handled when searching for media.
data FileType = Media FilePath | Zip FilePath | Unsupported
              deriving Show

main :: IO ()
main = getArgs >>= execute . parseArgs
  where
    parseArgs ("add":xs) = if (not . null) xs then Add xs else Invalid
    parseArgs ("help":_) = Help
    parseArgs (x:_) = Unknown x
    parseArgs _     = Invalid

--- Print program usage to stdout.
showUsage :: IO ()
showUsage =
  putStrLn $ unlines
  [ "Usage:"
  , "  add [items...]   Add files or folders to the iTunes library"
  , "  help             Show usage" ]

--- Run the program as specified by the program arguments.
execute :: Args -> IO ()
execute Help          = putStrLn "itunes: Commands for working with iTunes" >> showUsage
execute Invalid       = putStrLn "Invalid usage." >> showUsage >> exitFailure
execute (Unknown cmd) = putStrLn ("Unrecognised command: " ++ cmd) >> showUsage >> exitFailure
execute (Add args)    = do
  itunesExists <- itunesMedia >>= doesDirectoryExist
  unless itunesExists $ putStrLn "Cannot find iTunes Media folder" >> exitFailure
  paths <- pathsFromArgs
  warnWhereNotExists paths
  media <- liftM concat $ mapM mediaFromPath paths
  when (null media) $ putStrLn "No media found." >> exitFailure
  mapM_ addToItunes media
  promptDeleteOriginals media

  where
    pathsFromArgs =
      forM args $ \path ->
        canonicalizePath path `catch` (\(_::IOException) -> return path)

    warnWhereNotExists paths = do
      notExists <- filterM (liftM not . fileOrDirectoryExists) paths
      unless (null notExists) $ do
        putDoc $ dullyellow (text "Warning: the following items do not exist:") <> linebreak
        mapM_ (\x -> putDoc $ dullyellow (text "  ? ") <+> text x <> linebreak) notExists

    promptDeleteOriginals files = do
      let n = length files
      putStrLn $ "Delete original " ++ pluralize n "item" ++ "? (y/n) [n] "
      shouldDelete <- getYesOrNo False
      when shouldDelete $ do
        forM_ files $ \x -> do
          removeFile x
          putDoc $ red (text "  D ") <+> text x <> linebreak

        putStrLn $ "Deleted " ++ show n ++ " " ++ pluralize n "item" ++ "."

--- Concatenate a monadic filepath with pure filepaths.
(/>) :: IO FilePath -> FilePath -> IO FilePath
io /> p = (</>) <$> io <*> pure p
infix 4 />

--- The path to the iTunes library in the user's home folder.
itunesMedia :: IO FilePath
itunesMedia = getHomeDirectory /> "Music" </> "iTunes" </> "iTunes Media"

--- Copy the given file to the iTunes library.
addToItunes :: FilePath -> IO ()
addToItunes file = do
  dest <- itunesMedia /> "Automatically Add to iTunes.localized" </> takeFileName file
  copyFile file dest
  putDoc $ green (text "  A ") <+> text (takeFileName file) <> linebreak

--- Valid media extensions
isMedia :: FilePath -> Bool
isMedia path = takeExtension path `elem` [".m4a", ".m4v", ".mov",
                                     ".mp4", ".mp3", ".mpg",
                                     ".aac", ".aiff"]

--- Search for media files under the given filepath.
mediaFromPath :: FilePath -> IO [FilePath]
mediaFromPath path = do
  exists <- fileOrDirectoryExists path
  if exists
    then liftM concat $ getFilesInTree path >>= mapM categoriseType >>= mapM selectMedia
    else return []

--- Inspect the given file and determine whether it is an actionable type.
categoriseType :: FilePath -> IO FileType
categoriseType p@(isMedia -> True) = return (Media p)
categoriseType p = do
  -- Read file header.
  bs <- liftM (L8.unpack . L8.take 2) (L8.readFile p)
  return $ case bs of
    "PK" -> Zip p
    _         ->  Unsupported

--- Map the given file to its media items. Search archives for media.
selectMedia :: FileType -> IO [FilePath]
selectMedia (Media m)   = return [m]
selectMedia Unsupported = return []
selectMedia (Zip _)     = undefined

--- Walk the directory tree to find all files below a given path.
getFilesInTree :: FilePath -> IO [FilePath]
getFilesInTree d | takeFileName d `elem` [".", ".."] = return []
getFilesInTree d = do
  isDir <- doesDirectoryExist d
  isFile <- doesFileExist d
  case (isDir, isFile) of
    (True, _) -> concat <$> (getDirectoryContents d >>= mapM (getFilesInTree . (</>) d))
    (_, True) -> return [d]
    _         -> return []

--- Test whether the given file or directory exists.
fileOrDirectoryExists :: FilePath -> IO Bool
fileOrDirectoryExists x = or <$> sequence [doesDirectoryExist x, doesFileExist x]

type Count = Int
--- Perform a naive string pluralisation.
pluralize :: Count -> String -> String
pluralize 1 str = str
pluralize _ str = str ++ "s"

type Default = Bool
--- Prompt the user for a yes or no response, with a default answer.
getYesOrNo :: Default -> IO Bool
getYesOrNo deflt = do
  ch <- getChar
  case toLower ch of
    'y'  -> return True
    'n'  -> return False
    '\n' -> return deflt
    _    -> getYesOrNo deflt

File: itunes.rb
#!/usr/bin/env ruby
#
# itunes
#
# AUTHOR: Chris Barrett <chris.d.barrett@me.com>
# LICENSE: BSD
#
# Copyright (c) 2013, Chris Barrett
#
#
# DESCRIPTION:
#
#  Commands for working with iTunes from the command-line.
#

require 'fileutils'
require 'io/console'
require 'pathname'
require 'colorize'

def print_usage
  puts <<-EOF
itunes: Commands for working with iTunes from the command-line.

Usage:
  add [items...]    Add files or folders to the iTunes library
  help              Show usage

EOF
end

# Find the iTunes Media folder. If it is not set in the environment,
# use the default location.
ITUNES_MEDIA = ENV['ITUNES_MEDIA'] || File.join(Dir.home,
                                                'Music',
                                                'iTunes',
                                                'iTunes Media')

# Prompt the user for yes-or-no input with a default value.
def read_y_or_n(default)
  case STDIN.getch.downcase
  when 'y'  then true
  when 'n'  then false
  when '\r', '\n' then default
  else
    puts 'Enter `y` or `n`.'.red
    read_y_or_n default
  end
end

class String
  def pluralize(n)
    n == 1 ? self : "#{self}s"
  end
end

# Traverse the directory tree, returning all the files in the hierarchy.
def files_in_tree(d)
  Dir.chdir(d) do
    Dir['**'].map do |f|
      f = File.join(d, f)
      File.directory?(f) ? files_in_tree(f) : f
    end
  end.flatten
end

# Add a collection of items to the iTunes library. Accepts both files and
# directories.
def add_to_itunes(xs)
  # Get all the files in the collection and sort by name.
  files = xs
    .map { |x| File.absolute_path x }
    .map { |x| File.directory?(x) ? files_in_tree(x) : x }
    .flatten
  # Copy files, showing progress.
  dest = File.join(ITUNES_MEDIA, 'Automatically Add to iTunes.localized')
  files.each do |f|
    FileUtils.cp(f, dest)
    puts " #{'+'.green} #{Pathname.new(f).basename}"
  end
  # Delete the original items if requested by the user.
  n = files.count
  puts "Delete original #{'item'.pluralize n}? (y/n) [n] "
  if read_y_or_n(false)
    n = xs.each do |x|
      FileUtils.rm_r x
      puts " #{'D'.red} #{x}"
    end.count
    puts "Deleted #{n} #{'item'.pluralize n}."
  end
end

# Run main.
if __FILE__ == $PROGRAM_NAME
  cmd, *xs = ARGV
  case cmd
  when 'help' || '-h' || '--help' then print_usage
  when 'add' then
    abort 'Cannot find iTunes Media folder' unless Dir.exists? ITUNES_MEDIA
    targets, noexist = xs.partition { |x| File.exists? x }
    if noexist.any?
      puts 'Warning: the following items do not exist:'.yellow
      noexist.each { |x| puts " #{'?'.yellow} #{x}" }
      puts
    end
    if targets.any?
      add_to_itunes targets
    else
      abort 'No items added.'.red
    end
  else
    if cmd then abort "Unknown command: #{cmd}".red
    else
      print_usage
    end
  end
end
